packages/validator/lib/validator.js

```diff
@@ -0,0 +1,424 @@
+                  requiredPath = requiredPath.replace('.', '');
+                }
+
+                let index = 0;
+                let indexMatches = paramName.match(/(?<=\[)[\d]+(?=])/);
+
+                if (indexMatches && indexMatches.length) {
+                  index = parseInt(indexMatches[0]);
+                }
+
+                if (!workingDir[paramName]) {
+                  valid = false;
+                  messages.push({ level: 'error', message: `Missing parameter ${paramPathPrefix ? `${paramPathPrefix}.` : ''}${currentDir}${(currentDir && requiredPath) ? '.' : ''}${requiredPath}`});
+                  break;
+                }
+
+                if (!nonRedundantWD[paramName]) {
+                  if (typeof workingDir === 'object') {
+                    nonRedundantWD[paramName] = Array.isArray(workingDir[paramName]) ? [] : {};
+                  } else {
+                    nonRedundantWD[paramName] = {};
+                  }
+                }
+
+                nonRedundantWD = nonRedundantWD[paramName];
+                workingDir = workingDir[paramName];
+
+                if (index) {
+                  if (!workingDir[index]) {
+                    valid = false;
+                    messages.push({ level: 'error', message: `Array out of bounds, attempted to access element on index ${index} in ${paramPathPrefix ? `${paramPathPrefix}.` : ''}${currentDir}`}, paramPathPrefix);
+                    break;
+                  }
+
+                  workingDir = workingDir[index];
+
+                  nonRedundantWD.push({});
+                  nonRedundantWD = nonRedundantWD[nonRedundantWD.size() - 1];
+                }
+              }
+            }
+          }
+        }
+
+        break;
+
+      case '__regexp':
+        if (typeof specs !== 'string' || !specs.match(new RegExp(specsStruct[key]))) {
+          let level = 'warning';
+
+          if (specsStruct['__level']) {
+            level = specsStruct['__level'];
+
+            if (level === 'error') {
+              valid = false;
+            }
+          }
+
+          messages.push({ level, message: `${paramPath} is not formatted correctly` });
+        }
+
+        break;
+
+      case '__keyRegexp':
+        for (const item of Object.keys(specs)) {
+          if (!item.match(new RegExp(specsStruct[key]))) {
+            messages.push({ level: 'error', message: `Key ${item} in ${paramPath}${paramPath ? '.' : ''}${item} is formatted incorrectly` });
+          }
+        }
+
+        break;
+
+      case '__maxSize':
+        if (specsStruct[key] < specs.length) {
+          messages.push({ level: 'error', message: `${paramPath} must contain ${specsStruct[key]} or less items` });
+          valid = false;
+        }
+
+        break;
+
+      case '__arrayItem':
+        if (!nonRedundantParams) {
+          nonRedundantParams = [];
+        }
+
+        for (let i = 0; i < specs.length; i++) {
+          nonRedundantParams.push({});
```

Comment:
This part was validating arrays, if its the place I think it is. `nonRedundantParams` structure must be the same as structure of specs, so at the end we can check for extra parameters. In this case its necessary to push empty object into the array, so that when validateSpecs is recursively called few lines later, there is something that we can pass as `nonRedundantParams`. I hope that makes sense, I added some comments into the code, perhaps its more clear now.